DOC: improve description of the `data` entry in `__array_interface__` (#21248)
numpy/numpy: doc/source/reference/arrays.interface.rst

       **Default**: ``[('', typestr)]``

   **data** (optional)
       A 2-tuple whose first argument is a signed integer (capable of
       holding a pointer to ``void`` such as the C/C++ type
       `intptr_t <https://en.cppreference.com/w/cpp/types/integer>`__)

>        that points to the data-area storing the array

Reviewer: Thanks.  I think the reference to "long integer" in the old text refers to Python 2 where there was an integer that was stored as "long" and a arbitrary precision "long integer" and on some platforms just a "long" would not have been enough.
I expect the clearest thing here will be to just remove the entire thing and say: "a Python integer representing the pointer to the data-area" maybe?  We can probably assume the user is familiar with casting integers to pointers.

Author: Thanks! Yes, we can do that as well.

After googling a bit on the issue (SO & blogs), it seems to be a common pitfall that people serialize into a too-small integer type (especially since they are [platform dependent](https://en.cppreference.com/w/cpp/language/types)). Usually things work on the platform tested, but once they switch platforms or the memory address moves to another location in address space in later use cases things start failing and get hard to debug. Thus, I am inclined to give a bit more hints on this, but will follow our guidance.

Reviewer: Ah OK, if that is a common pitfall, then I am +1 on noting it.  NumPy actually just uses Python integers here (thus the old text), but of course any integer type is OK, so long it doesn't lose precision.

I would still prefer to start of with just saying `integer` and then maybe follow with: If you do not use Python's arbitrary precision integers, make sure that the integer type is large enough to hold a pointer (e.g. `intptr_t` compatible)?

Reviewer 2: I would go with "A 2-tuple whose first argument is the address of the data-area storing the array contents This must be the address of the first element of the data (in other ..." and avoid specifying the type.

Author: Hi mattip, I think not specifying the type as integer won't work. I tried storing this as hex memory address, which does not fulfill the protocol.

Reviewer: @ax3l you at some point had `long(pointer)` in the code.  That is an incorrect cast in C/C++ (on 64bit Windows at least).  

But that has little to do with `__array_interface__` itself since that bad cast happened before any Python/NumPy involvement.

Helping someone in your position seems good.  I like it to be clear that this is a tip/hint and not central to the `__array_interface__` definition.  Maybe something like: "For this, the pointer needs to be safely converted to a Python integer using `PyLong_FromVoidPtr` or via a cast to `uintptr_t` when using automatic wrappers like `pybind11`." 

(Note that technically, I believe `uintptr_t` is strictly more correct, although I doubt that it matters for any practical purposes.

Author: > @ax3l you at some point had long(pointer) in the code. That is an incorrect cast in C/C++ (on 64bit Windows at least).

Yes, that's how I spotted it. Worked on Linux and macOS but somehow started to fail on Windows... then realized `int` and `long` are too small on Windows 64bit (`LLP64` or `4/4/8` - `int` and `long` are 32-bit, pointer is 64-bit) and only `long long` covers 64bit.

Yes, I am also ok to maybe make it a `.. note::` box or so as a hint? `PyLong_FromVoidPtr` is a good hint as well :+1:

Reviewer: I would be happy with a "note", if you think it is likely to help others.

> unsigned (uintptr_t) does not cover the array interface protocol 

Very sure it needs to be unsigned at that step.  IIRC, `intptr_t` does not guarantee to represent all pointers as a positive integer.  So it is incorrect, since it has to be passed as positive integer in Python.
Note that for you `pybind11` will probably convert that integer you cast to to an arbitrary precision Python integer (or something compatible) for you.  So the type really doesn't matter (pybind11 takes care), but if `(intptr_t)pointer` is negative, it will pass it as a negative integer, which is incorrect.

Reviewer: OK... NumPy uses `PyLong_AsVoidPtr` which explicitly supports the negative values that `intptr_t` may get.  Now NumPy will never *produce* them (`PyLong_FromVoidPtr` does not produce negative values) so I think you should still prefer `uintptr_t`.  But the value you get by casting to `intptr_t` *is* effectively supported, at least by NumPy.

Author: Thanks for looking into this. I found contradicting information in C implementations, explicitly using signed types.

Where is saw signed values was here:
- `cupy`: `__cuda_array_interface__`, which is compatible in the `data` argument to `__array_interface__`
  - https://github.com/cupy/cupy/issues/4644#issuecomment-775480279
  - https://github.com/cupy/cupy/blob/a5b24f91d4d77fa03e6a4dd2ac954ff9a04e21f4/cupy/core/core.pyx#L2478-L2514
  - https://github.com/cupy/cupy/blob/4469fae998df33c72ff40ef954cb08b8f0004b18/cupy/cuda/memory.pxd
- `numba`: (same as above)
  - explicit `int` or `long` mention in https://numba.readthedocs.io/en/latest/cuda/cuda_array_interface.html
  - https://github.com/numba/numba/blob/637a75023bae889467ebfc0e8489be547b62a209/numba/cuda/cudadrv/devicearray.py#L123
  - https://github.com/numba/numba/blob/637a75023bae889467ebfc0e8489be547b62a209/numba/cuda/cudadrv/devicearray.py#L123

Hope this does not add to much to the confusion. But seems consistent with the `PyLong_AsVoidPtr` you found?

Reviewer 3: Cython handles casting back and forth between intptr_t/uintptr_t and Python int correctly for us (yes under the hood PyLong_AsVoidPtr would be used), so it's free to choose either signed or unsigned.

(Historically CuPy uses the former unconditionally for any data pointer, but with hindsight I think it only makes sense here (and a few other places) where pointer arithmetics might be needed/performed.)

Author: Thanks for the background, much appreciated :)

Since `__array_interface__` is a pure Python standard anyway (no stable C equivalent, if I understand it right), maybe the least confusing could be if we write ["a Python integer"](https://docs.python.org/3/c-api/long.html)?
```suggestion
       A 2-tuple whose first argument is a
       `Python integer <https://docs.python.org/3/c-api/long.html>`__
       that points to the data-area storing the array
```

And then we could write a `.. hint::` box such as "Note: When converting from C/C++ via `PyLong_From*` or high-level bindings such as Cython and pybind11, make sure to use an integer of sufficiently large bitness."

Reviewer: Sorry, this has been a while.  I think your last thought was good (not sure there is a `.. hint::` or whether it would just be a `.. note::` (I guess there is a `.. admonition:: Hint` that gives a title though.

Would you be up for doing that last change?  Otherwise, I guess whoever stumbles accross this next time, should just go ahead, do tweak a bit and merge.
